internal/server/polling/server.go

```diff
@@ -83,16 +97,20 @@ func (s *Server) poll(ctx context.Context, resource types.NamespacedName, secret
 		return fmt.Errorf("failed to get Source object: %w", err)
 	}
 
-	gitProvider, repo, err := provider.FromURL(
-		source.Spec.URL,
-		provider.WithLogger(s.log),
-		provider.WithToken("api-token", string(secret.Data["token"])),
-	)
+	fromURL := s.gitProviderFunc
+	if fromURL == nil {
+		fromURL = func(url string) (prLister, provider.Repository, error) {
+			return provider.FromURL(url,
+				provider.WithLogger(s.log),
+				provider.WithToken("api-token", string(secret.Data["token"])))
+		}
+	}
+	lister, repo, err := fromURL(source.Spec.URL)
 	if err != nil {
 		return fmt.Errorf("failed to get git provider: %w", err)
 	}
 
-	prs, err := gitProvider.ListPullRequests(ctx, repo)
+	prs, err := lister.ListPullRequests(ctx, repo)
```

Comment:
> We are using it to list pull requests and comments, but that does not mean it's a lister

Actually it does! The requirement for that argument is something that can list pull requests -- a lister. It doesn't matter what the concrete value thinks it is (see its use in the test).
That said, I don't really mind the name it's given, perhaps it could be `pullRequestProvider` or something like that.

Reply:
I eventually realised we don't have to worry about this -- the new revision just changes the method-under-test to `reconcile`, which is handed all its inputs rather than fetching them.